docs/content/setup/configuration.md
# Configuration

The following table lists all the configuration values that can be set as environment variables (e.g. via a `.env` file).

!!! note
    - Any changes to the configuration values listed below require an application restart to take effect.
    - Some variable names contain _double_ underscores (`__`) to reflect the nested internal configuration – be careful not to miss those in case you're into typing things by hand!


## Core Config

Basic configuration, URL, paths, ...

### `TEKST_SERVER_URL`
Full public URL of the server running the application (without sub path, port only if other than 80/443) (String – default: `http://127.0.0.1:8000`)

### `TEKST_API_PATH`
Public sub path of the API. This **must** be set independently from `TEKST_WEB_PATH`, so if `TEKST_WEB_PATH` is `/foo`, `TEKST_API_PATH` must include the web path, like `/foo/api`. (String – default: `/api`)

!!! info
    The official docker image **assumes an API path ending with `/api`**, so either **do not** set `TEKST_API_PATH` at all (if you deploy Tekst under `/`) to let it default to `/api`, or set an API path that is a combination of `TEKST_WEB_PATH` and the path suffix `/api`, e.g. `/foo/api`!

### `TEKST_WEB_PATH`
Public path of the web client (String – default: `/`)

### `TEKST_BEHIND_REVERSE_PROXY`
If the API runs behind a reverse proxy (e.g. if using the the project's `compose.yml`), set this to `true` and make sure the reverse proxy is properly setting (or appending to) the [`X-Forwarded-For`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) header on the passed requests. This is to make sure the API is able to **try to** prevent accidentally flooding the workers with long-running background jobs. (Boolean – default: `false`)

### `TEKST_WEB_STATIC_DIR`
Local path to static web files directory (String – default: `/var/www/tekst/static/`)

### `TEKST_DEV_MODE`
Development mode – only use for development! (Boolean – default: `false`)

### `TEKST_LOG_LEVEL`
API log level (`DEBUG` | `INFO` | `WARNING` | `ERROR` | `CRITICAL` – default: `WARNING`)

### `TEKST_AUTO_MIGRATE`
If `false` (default) and there are pending DB migrations that should be run, the bootstrap routine will fail with a critical log message (as the API itself will always do). If `true`, pending migrations will be run as part of the bootstrap routine. (Boolean – default: `false`)

### `TEKST_XSRF`
Enables/disables [XSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery) protection enforced by the API. (Boolean – default: `true`)

### `TEKST_TEMP_FILES_DIR`
Absolute path to local temporary directory to use (String – default: `/tmp/tekst_tmp`)



## Database (MongoDB)

Configuration for the MongoDB connection

### `TEKST_DB__PROTOCOL`
MongoDB protocol (String – default: `mongodb`)

### `TEKST_DB__HOST`
MongoDB host (String – default: `127.0.0.1`)

### `TEKST_DB__PORT`
MongoDB port (Integer – default: `27017`)

### `TEKST_DB__USER`
MongoDB user (String – default: _none_)

### `TEKST_DB__PASSWORD`
MongoDB password (String – default: _none_)

### `TEKST_DB__NAME`
MongoDB database name (String – default: `tekst`)

### `TEKST_DB__UNICODE_NF`
Apply Unicode string normalization when writing to the database (`NFC`, `NFKC`, `NFD`, `NFKD` or _none_ – default: `NFC`)



## Search Server (Elasticsearch)

Configuration for the connection to the Elasticsearch server

### `TEKST_ES__PROTOCOL`
Elasticsearch protocol (String – default: `http`)

### `TEKST_ES__HOST`
Elasticsearch host (String – default: `127.0.0.1`)

### `TEKST_ES__PORT`
Elasticsearch port (Integer – default: `9200`)

### `TEKST_ES__PREFIX`
Elasticsearch prefix (for index, templates, etc.) (String – default: `tekst`)

### `TEKST_ES__TIMEOUT_INIT_S`
Timeout for waiting for Elasticsearch service to be available on startup, in seconds (Integer – default: `240`)

### `TEKST_ES__TIMEOUT_GENERAL_S`
General client timeout for requests to Elasticsearch, in seconds (Integer – default: `30`)

### `TEKST_ES__TIMEOUT_SEARCH_S`
Timeout for search reqests to Elasticsearch, in seconds (Integer – default: `30`)



## Security

All the configuration related to app-level security

### `TEKST_SECURITY__SECRET`
Secret to use for token generation (String – default: `must_change_this`)

### `TEKST_SECURITY__CLOSED_MODE`
Disallow public user signups/applications; goto `/login` to log in as an existing user (Boolean – default: `false`)

### `TEKST_SECURITY__USERS_ACTIVE_BY_DEFAULT`
Whether new user accounts are active by default (otherwise have to be activated by admins) (Boolean – default: `false`)

### `TEKST_SECURITY__ENABLE_COOKIE_AUTH`
Enable cookie-based authentication (needed for the web client to work!) (Boolean – default: `true`)

### `TEKST_SECURITY__AUTH_COOKIE_NAME`
Cookie name (String – default: `tekstuserauth`)

### `TEKST_SECURITY__AUTH_COOKIE_DOMAIN`
Cookie domain (String – default: _none_)

### `TEKST_SECURITY__AUTH_COOKIE_LIFETIME`
Cookie lifetime in seconds – `0` produces session cookies (Integer – default: `43200`)

### `TEKST_SECURITY__ACCESS_TOKEN_LIFETIME`
Lifetime of access tokens in seconds; limits validity of cookies and must thus be greater than or equal to `TEKST_SECURITY__AUTH_COOKIE_LIFETIME` (Integer – default: `43200`, min.: `3600`)

### `TEKST_SECURITY__ENABLE_JWT_AUTH`
Enable JWT-based authentication (optional, useful for programmatic access to the API) (Boolean – default: `false`)

### `TEKST_SECURITY__AUTH_JWT_LIFETIME`
Lifetime of JWT used for authentication (Integer – default: `86400`)

### `TEKST_SECURITY__RESET_PW_TOKEN_LIFETIME`
Lifetime of password reset tokens (Integer – default: `3600`)

### `TEKST_SECURITY__VERIFICATION_TOKEN_LIFETIME`
Lifetime of account verification tokens (Integer – default: `86400`)

### `TEKST_SECURITY__INIT_ADMIN_EMAIL`
Email address of initial admin account (String – default: _none_)

### `TEKST_SECURITY__INIT_ADMIN_PASSWORD`
Password of initial admin account (String – default: _none_)



## Email

Configuration for the SMTP connection and sender address of outgoing emails

### `TEKST_EMAIL__SMTP_SERVER`
SMTP server address (String – default: `127.0.0.1`)

### `TEKST_EMAIL__SMTP_PORT`
SMTP server port (Integer – default: `25`)

### `TEKST_EMAIL__SMTP_USER`
SMTP user (String – default: _none_)

### `TEKST_EMAIL__SMTP_PASSWORD`
SMTP password (String – default: _none_)

### `TEKST_EMAIL__SMTP_STARTTLS`
Whether to use StartTLS for SMTP connection (Boolean – default: `true`)

### `TEKST_EMAIL__FROM_ADDRESS`
From-address used for outgoing emails (String – default: `noreply@example-tekst-instance.org`)



## Documentation

Configuration for the OpenAPI schema URL and API documentation via SwaggerUI and/or Redoc

### `TEKST_API__OPENAPI_URL`
URL sub path (under `TEKST_API_PATH`) to public `openapi.json` API schema (if not set, disables SwaggerUI and Redoc) (String – default: `/openapi.json`)

### `TEKST_API__SWAGGERUI_URL`
URL sub path (under `TEKST_API_PATH`) to SwaggerUI interactive API docs; empty value disables SwaggerUI (String – default: `/docs`)

### `TEKST_API__REDOC_URL`
URL sub path (under `TEKST_API_PATH`) to Redoc API docs; empty value disables Redoc (String – default: `/redoc`)

### `TEKST_API__TITLE`
Platform API title (String – default: `Tekst`)

### `TEKST_API__SUMMARY`
Platform API summary (String – default: _none_)

### `TEKST_API__DESCRIPTION`
Platform API description (String – default: _none_)

### `TEKST_API__TERMS_URL`
URL to API usage terms (String – default: _none_)

### `TEKST_API__CONTACT_NAME`
API contact name (String – default: _none_)

### `TEKST_API__CONTACT_EMAIL`
API contact email (String – default: _none_)

### `TEKST_API__CONTACT_URL`
API contact URL (String – default: _none_)

### `TEKST_API__LICENSE_NAME`
API license name; if not set, no license info will be integrated in API docs (String – default: _none_)

### `TEKST_API__LICENSE_ID`
API license ID (canonical short form of name) (String – default: _none_)

### `TEKST_API__LICENSE_URL`
API license URL (String – default: _none_)



## Cross-Origin Resource Sharing (CORS)

You _should_ handle CORS via your webserver configuration. But if for some reason you nead app-level CORS management, this is where to look.

### `TEKST_CORS__ENABLE`
Enable CORS header control; only enable this if you don't have a web server / reverse proxy that already handles it! (Boolean – default: `false`)

### `TEKST_CORS__ALLOW_ORIGINS`
CORS allow origins (Comma-separated list as string – default: `*`)

### `TEKST_CORS__ALLOW_CREDENTIALS`
CORS allow credentials (Boolean – default: `true`)

### `TEKST_CORS__ALLOW_METHODS`
CORS allow methods (Comma-separated list as string – default: `*`)

### `TEKST_CORS__ALLOW_HEADERS`
CORS allow headers (Comma-separated list as string – default: `*`)



## What is left

All these configuration values – and then some...

### `TEKST_MISC__USRMSG_FORCE_DELETE_AFTER_DAYS`
Delete old user messages after n days (Integer – default: `365`)

### `TEKST_MISC__MAX_RESOURCES_PER_USER`
Maximum number of resources/versions that one user is allowed to own (Integer – default: `10`)

### `TEKST_MISC__DEL_EXPORTS_AFTER_MINUTES`
Time in minutes after whichfinished/failed tasks that produce a downloadable file artifact (namely "exports") will be deleted automatically, including the respective file (Integer – default: `5`)


Tekst-Web/i18n/help/deDE/contentCssConfig.u.md
# Abweichendes CSS

Hier können Sie eigene CSS-Eigenschaften für Inhalte dieser Ressource festlegen. Bitte beachten Sie, dass sich alle angewendeten Regeln auf das gesamte Containerelement der Inhalte auswirken.


Tekst-API/tekst/config.py
import os
import re

from functools import cache
from os.path import realpath
from pathlib import Path
from secrets import token_hex
from typing import Annotated, Any, Literal
from urllib.parse import quote
from uuid import uuid4

from fastapi import Depends
from humps import camelize
from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    DirectoryPath,
    EmailStr,
    Field,
    StringConstraints,
    computed_field,
    field_validator,
)
from pydantic_settings import BaseSettings, SettingsConfigDict

from tekst import package_metadata
from tekst.types import FalsyToNone, HttpUrl, MultiLineString, SingleLineString


_DEV_MODE: bool = bool(os.environ.get("TEKST_DEV_MODE", False))


class ConfigSubSection(BaseModel):
    """Base class for config sub sections"""

    model_config = ConfigDict(
        alias_generator=camelize,
        validate_by_name=True,
        validate_by_alias=True,
        from_attributes=True,
        validate_assignment=True,  # should only happen in tests anyway
    )


def _select_env_files() -> list[str]:
    """
    Selects the dotenv (.env) files to load.
    Values from these files override the default
    values used in the config model classes.

    Selection and priority work as follows:

    - A maximum of three env files will be loaded
      (".env", one of ".env.dev" and ".env.prod",
      and, if defined, an additional custom env file)
    - Env files loaded later override values of env files loaded earlier

    Selection steps:

    1. ".env" if it exists
    2. ".env.dev" if it exists AND if TEKST_DEV_MODE env var is set to `true`
    3. ".env.prod" if it exists AND if TEKST_DEV_MODE env var is set to `false`
    4. Custom env file defined in "TEKST_CUSTOM_ENV_FILE" env var if it exists

    :return: List of .env file paths
    :rtype: list[str]
    """
    env_files = []
    # define used env file names
    f_env = ".env"
    f_env_dev = ".env.dev"
    f_env_prod = ".env.prod"
    f_env_custom = os.environ.get("TEKST_CUSTOM_ENV_FILE")
    # prio logic
    if os.path.exists(f_env):  # pragma: no cover
        env_files.append(f_env)
    if _DEV_MODE and os.path.exists(f_env_dev):
        env_files.append(f_env_dev)
    elif os.path.exists(f_env_prod):  # pragma: no cover
        env_files.append(f_env_prod)
    if f_env_custom and os.path.exists(f_env_custom):  # pragma: no cover
        env_files.append(f_env_custom)
    return env_files


class MongoDBConfig(ConfigSubSection):
    """Database config sub section model"""

    protocol: str = "mongodb"
    host: str = "127.0.0.1"
    port: int = 27017
    user: Annotated[
        str | None,
        StringConstraints(max_length=64),
        SingleLineString,
        FalsyToNone,
    ] = None
    password: Annotated[
        str | None,
        StringConstraints(max_length=64),
        SingleLineString,
        FalsyToNone,
    ] = None
    name: str = "tekst"
    unicode_nf: Literal["NFC", "NFKC", "NFD", "NFKD"] | None = "NFC"

    @field_validator("name", mode="before")
    @classmethod
    def validate_db_name(cls, v: Any) -> str:
        if not isinstance(v, str):
            v = str(v)
        if not re.match(r"^[a-zA-Z0-9_]+$", v):
            raise ValueError(f"Invalid database name: {v} (only [a-zA-Z0-9_] allowed)")
        return v

    @field_validator("host", "password", mode="before")
    @classmethod
    def url_quote(cls, v) -> str:
        if v is None:
            return v
        return quote(str(v).encode("utf8"), safe="")

    @computed_field
    @property
    def uri(self) -> str:
        creds = f"{self.user}:{self.password}@" if self.user and self.password else ""
        return f"{self.protocol}://{creds}{self.host}:{str(self.port)}"


class ElasticsearchConfig(ConfigSubSection):
    """Elasticsearch config sub section model"""

    protocol: str = "http"
    host: str = "127.0.0.1"
    port: int = 9200
    prefix: str = "tekst"
    timeout_init_s: int = 240
    timeout_general_s: int = 30
    timeout_search_s: str = "30s"

    @field_validator("host", mode="before")
    @classmethod
    def url_quote(cls, v) -> str:
        if v is None:
            return v
        return quote(str(v).encode("utf8"), safe="")

    @field_validator("timeout_search_s", mode="before")
    @classmethod
    def timeout_int_to_time_value(cls, v) -> str:
        if isinstance(v, int):
            return f"{v}s"
        v = re.sub(r"\D", "", v)  # keep only digits
        if not v:
            return "30s"
        return f"{v}s"

    @computed_field
    @property
    def uri(self) -> str:
        return f"{self.protocol}://{self.host}:{str(self.port)}"


class SecurityConfig(ConfigSubSection):
    """Security config sub section model"""

    secret: Annotated[
        str,
        StringConstraints(
            min_length=16,
            max_length=512,
        ),
        SingleLineString,
    ] = token_hex(32)
    closed_mode: bool = False
    users_active_by_default: bool = False

    enable_cookie_auth: bool = True
    auth_cookie_name: str = "tekstuserauth"
    auth_cookie_domain: Annotated[
        str | None,
        StringConstraints(max_length=64),
        SingleLineString,
        FalsyToNone,
    ] = None
    auth_cookie_lifetime: Annotated[
        int,
        Field(ge=3600),
    ] = 43200
    access_token_lifetime: Annotated[
        int,
        Field(ge=3600),
    ] = 43200

    enable_jwt_auth: bool = False
    auth_jwt_lifetime: Annotated[
        int,
        Field(ge=3600),
    ] = 86400

    reset_pw_token_lifetime: Annotated[
        int,
        Field(ge=600),
    ] = 3600  # 1h
    verification_token_lifetime: Annotated[
        int,
        Field(ge=600),
    ] = 86400  # 24h

    init_admin_email: Annotated[
        str | None,
        StringConstraints(max_length=256),
        SingleLineString,
        FalsyToNone,
    ] = None
    init_admin_password: Annotated[
        str | None,
        StringConstraints(max_length=256),
        SingleLineString,
        FalsyToNone,
    ] = None


class EMailConfig(ConfigSubSection):
    """Email-related things config sub section model"""

    smtp_server: Annotated[
        str,
        StringConstraints(max_length=256),
        SingleLineString,
    ] = "127.0.0.1"
    smtp_port: int = 25
    smtp_user: Annotated[
        str | None,
        StringConstraints(max_length=256),
        SingleLineString,
        FalsyToNone,
    ] = None
    smtp_password: Annotated[
        str | None,
        StringConstraints(max_length=256),
        SingleLineString,
        FalsyToNone,
    ] = None
    smtp_starttls: bool = True
    from_address: str = "noreply@example-tekst-instance.org"


class ApiDocConfig(ConfigSubSection):
    """Documentation config sub section model"""

    openapi_url: str = "/openapi.json"
    swaggerui_url: str = "/docs"
    redoc_url: str = "/redoc"

    title: Annotated[
        str,
        StringConstraints(max_length=32),
        SingleLineString,
    ] = "Tekst"
    summary: Annotated[
        str | None,
        StringConstraints(max_length=256),
        SingleLineString,
        FalsyToNone,
    ] = None
    description: Annotated[
        str | None,
        StringConstraints(max_length=4096),
        MultiLineString,
        FalsyToNone,
    ] = None
    terms_url: Annotated[HttpUrl | None, FalsyToNone] = None
    contact_name: Annotated[
        str | None,
        StringConstraints(max_length=64),
        SingleLineString,
        FalsyToNone,
    ] = None
    contact_email: Annotated[
        EmailStr | None,
        FalsyToNone,
    ] = None
    contact_url: Annotated[HttpUrl | None, FalsyToNone] = None
    license_name: Annotated[
        str | None,
        StringConstraints(max_length=32),
        SingleLineString,
        FalsyToNone,
    ] = None
    license_id: Annotated[
        str | None,
        StringConstraints(max_length=32),
        SingleLineString,
        FalsyToNone,
    ] = None
    license_url: Annotated[HttpUrl | None, FalsyToNone] = None


class CORSConfig(ConfigSubSection):
    """CORS config sub section model"""

    enable: bool = False
    allow_origins: str | list[str] = ["*"]
    allow_credentials: bool = True
    allow_methods: str | list[str] = ["*"]
    allow_headers: str | list[str] = ["*"]

    @field_validator("allow_origins", "allow_methods", "allow_headers", mode="after")
    @classmethod
    def split_cors(cls, v):
        if isinstance(v, str):
            return [e.strip() for e in v.split(",")]
        return v


class MiscConfig(ConfigSubSection):
    """Misc config sub section model"""

    usrmsg_force_delete_after_days: int = 365
    max_resources_per_user: int = 10
    del_exports_after_minutes: int = 5

    @computed_field
    @property
    def demo_data_path(self) -> Path:
        return Path(realpath(__file__)).parent.parent / "demo"


class TekstConfig(BaseSettings):
    """Platform config model"""

    model_config = SettingsConfigDict(
        env_file=_select_env_files(),
        env_file_encoding="utf-8",
        env_prefix="TEKST_",
        env_nested_delimiter="__",
        validate_by_name=True,
        validate_by_alias=True,
        case_sensitive=False,
        protected_namespaces=("model_",),
        validate_assignment=True,  # should only happen in tests anyway
    )

    server_url: HttpUrl = "http://127.0.0.1:8000"
    api_path: str = "/api"
    web_path: str = "/"
    behind_reverse_proxy: bool = False

    dev_mode: bool = False
    log_level: str = "warning"
    auto_migrate: bool = False
    xsrf: bool = True

    temp_files_dir: Annotated[
        DirectoryPath,
        BeforeValidator(lambda v: Path(v) if isinstance(v, str) else v),
    ] = Path("/tmp/tekst_tmp")

    # config sub sections
    db: MongoDBConfig = MongoDBConfig()  # MongoDB-related config
    es: ElasticsearchConfig = ElasticsearchConfig()  # Elasticsearch-related config
    security: SecurityConfig = SecurityConfig()  # security-related config
    email: EMailConfig = EMailConfig()  # Email-related config
    api_doc: ApiDocConfig = ApiDocConfig()  # API documentation-related config
    cors: CORSConfig = CORSConfig()  # CORS-related config
    misc: MiscConfig = MiscConfig()  # misc config

    @computed_field
    @property
    def tekst(self) -> dict[str, str]:
        return dict(name="Tekst", **package_metadata)

    @field_validator("dev_mode", mode="before")
    @classmethod
    def parse_dev_mode(cls, v: Any) -> bool:
        return False if str(v).lower() == "false" else bool(v)

    @field_validator("log_level", mode="after")
    @classmethod
    def uppercase_log_lvl(cls, v: str) -> str:
        return v.upper()

    @field_validator("temp_files_dir", mode="before")
    @classmethod
    def temp_dir_to_existing_path_obj(cls, v: Any) -> Path:
        try:
            path = Path(str(v))
            path.mkdir(parents=True, exist_ok=True)
            test_file_path = path / (str(uuid4()) + ".tmp")
            test_file_path.unlink(missing_ok=True)
            test_file_path.write_text(data="test")
            test_file_path.unlink(missing_ok=True)
        except Exception as e:
            print(e)
            raise ValueError(
                f"Temporary directoy is not valid or writable: {str(v)}"
            ) from e
        return path


@cache
def get_config() -> TekstConfig:
    return TekstConfig()


ConfigDep = Annotated[TekstConfig, Depends(get_config)]
